h2o-core/src/main/java/water/fvec/NewChunk.java

```diff
@@ -491,6 +501,46 @@ public void addNA() {
     ++_len;
   }
 
+  public void addNumDecompose(final double d) {
+    if (isUUID() || isString()) { // not worth trying
+      addNA();
+      return;
+    }
+    if (_ds != null) { // already using doubles
+      addNum(d);
+      return;
+    }
+    if ((long) d == d) {
+      addNum((long) d, 0);
+      return;
+    }
+
+    final int expIdx = Math.getExponent(d) - Double.MIN_EXPONENT;
+    if (expIdx == -1) { // zero or subnormal
+      if (d == 0)
+        addNum(0, 0);
+      else // subnormal
+        addNum(d);
+      return;
+    }
+
+    final int sign = d < 0 ? -1 : 1;
+    int exp = EXP10s[expIdx];
+    double val = sign * d * INV_POW10s[expIdx];
+
+    while ((long) val != val) {
```

Comment:
This is fast and precise, yet there is a way for us to be more precise.

Precision comparison code:
```java
        //Iteration conversion
        double val = 0.92233720368547758079223372036854775807;
        int loopPrecision = 0;
        while ((long) val != val) {
            double x = val * 10;
            if (x > Long.MAX_VALUE) {
                System.out.println("Broken " + loopPrecision);
                break;
            }
            val = x;
            loopPrecision++;
        }

        System.out.println("Iteration precision " + loopPrecision);
        System.out.println("Iteration value " + val);

        System.out.println("----------------");

        //BigDecimal conversion
        BigDecimal bigDecimal = new BigDecimal(0.9223372036854775807, MathContext.DECIMAL64);
        final int decimalPoints = Math.max(0, bigDecimal.stripTrailingZeros().scale());
        System.out.println("Big Decimal precision: " + decimalPoints);
        System.out.println("Big Decimal value: " + bigDecimal.movePointRight(decimalPoints));
```

## Code output

```plain
/home/pavel/prg/jdk1.8.0_171/bin/java -javaagent:/home/pavel/prg/idea-IU-181.5281.24/lib/idea_rt.jar=36873:/home/pavel/prg/idea-IU-181.5281.24/bin -Dfile.encoding=UTF-8 -classpath /home/pavel/prg/jdk1.8.0_171/jre/lib/charsets.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/deploy.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/cldrdata.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/dnsns.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/jaccess.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/jfxrt.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/localedata.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/nashorn.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/sunec.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/sunjce_provider.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/sunpkcs11.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/ext/zipfs.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/javaws.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/jce.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/jfr.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/jfxswt.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/jsse.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/management-agent.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/plugin.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/resources.jar:/home/pavel/prg/jdk1.8.0_171/jre/lib/rt.jar:/home/pavel/wrk/test/out/production/classes Main
Iteration precision 15
Iteration value 9.22337203685478E14
----------------
Big Decimal precision: 16
Big Decimal value: 9223372036854776

Process finished with exit code 0
```

## Performance
Performance of the second solution is a question. It is very slow, approx. 23-25x slower than your solution.

```java
public static void main(String[] args) {

        Random random = new Random(0);
        final int nSamples = 1000000;
        final double val = 0.92233720368547758079223372036854775807;

        //Warmup
        for (int i = 0; i < nSamples; i++) {
            iteration(val);
        }

        long startTime = System.currentTimeMillis();
        for (int i = 0; i < nSamples; i++) {
            iteration(val);
        }
        long endTime = System.currentTimeMillis();
        System.out.println("Iteration time: " + (endTime - startTime));

        //////////////////// BigDecimal

        //Warmup BigDecimal
        for (int i = 0; i < nSamples; i++) {
            bigDecimal(val);
        }

        startTime = System.currentTimeMillis();
        for (int i = 0; i < nSamples; i++) {
            bigDecimal(val);
        }
        endTime = System.currentTimeMillis();
        System.out.println("BigDecimal time: " + (endTime - startTime));


    }

    private static void iteration(double val){
        int loopPrecision = 0;
        while ((long) val != val) {
            double x = val * 10;
            if (x > Long.MAX_VALUE) {
                break;
            }
            val = x;
            loopPrecision++;
        }
    }

    private static void bigDecimal(final double val){
        BigDecimal bigDecimal = new BigDecimal(val, MathContext.DECIMAL64);
        bigDecimal.movePointRight(Math.max(0, bigDecimal.scale()));
    }
```

Output:

```plain
Iteration time: 20
BigDecimal time: 537
```

Even if you scale the number of samples, the ratio remains more or less the same.

Reply:
Yes, this is a good point and a known issue - PUBDEV-5840. My goal was to keep the compatibility with the previous version. We will improve this implementation further in PUBDEV-5840.